grr/server/grr_response_server/bin/grrafana.py

```diff
@@ -0,0 +1,165 @@
+    """Given a client ID as a Grafana variable and targets (resource usages),
+    returns datapoints in a format Grafana can interpret."""
+    request_json_format = request.json
+    requested_client_id = extractClientIdFromVariable(request_json_format)  # There must be a ClientID variable declated in Grafana.
+    requested_targets = [entry["target"] for entry in request_json_format["targets"]]
+    response = fetchDatapointsForTargets(requested_client_id, request_json_format["maxDataPoints"], requested_targets)
+    return JSONResponse(response=json.dumps(response), mimetype="application/json")
+
+  def onAnnotations(self, request: JSONRequest) -> JSONResponse:
+    pass
+
+
+def fetchClientIds() -> List[Text]:
+  """Fetches GRR client IDs that have resource usage records in Fleetspeak database."""
+  return fleetspeak_utils.GetClientIdsFromFleetspeak()
+
+
+Datapoint = Tuple[float, int]
+Datapoints = List[Datapoint]
+TargetWithDatapoints = Dict[Text, Datapoints]
+def fetchDatapointsForTargets(client_id: Text, limit: int, targets: List[Text]) -> List[TargetWithDatapoints]:
+  """Fetches a list of <datapoint, timestamp> tuples for each target metric from Fleetspeak database."""
+  records_list = fleetspeak_utils.FetchClientResourceUsageRecordsFromFleetspeak(client_id, limit)
+  response = list()
+  for target in targets:
+    datapoints_for_single_target = createDatapointsForTarget(target, records_list)
+    target_datapoints_dict = typing.cast(TargetWithDatapoints, {"target": target, "datapoints": datapoints_for_single_target})
+    response.append(target_datapoints_dict)
+  return response
+
+
+def createDatapointsForTarget(target: Text, records_list: List[ClientResourceUsageRecord]) -> Datapoints:
+  if target == "mean_user_cpu_rate":
+    record_values = [record.mean_user_cpu_rate for record in records_list]
+  elif target == "max_user_cpu_rate":
+    record_values = [record.max_user_cpu_rate for record in records_list]
+  elif target == "mean_system_cpu_rate":
+    record_values = [record.mean_system_cpu_rate for record in records_list]
+  elif target == "max_system_cpu_rate":
+    record_values = [record.max_system_cpu_rate for record in records_list]
+  elif target == "mean_resident_memory_mib":
+    record_values = [record.mean_resident_memory_mib for record in records_list]
+  elif target == "max_resident_memory_mib":
+    record_values = [record.max_resident_memory_mib for record in records_list]
+  return [(v, r.server_timestamp.seconds * 1000) for (v, r) in zip(record_values, records_list)]
+
+
+def extractClientIdFromVariable(req: JSONRequest) -> Text:
+  """Extracts the client ID from a Grafana JSON request."""
+  # Based on an assumption that there is only one Grafana variable.
+  return next(iter(req["scopedVars"].values()))["value"]
+
+
+def main(argv: Any) -> None:
+  """Main."""
+  del argv  # Unused.
+
+  if flags.FLAGS.version:
+    print("GRRafana server {}".format(config_server.VERSION["packageversion"]))
```

Comment:
Prefer to use f-strings (`f"GRRafana server {config_server.VERSION["packageversion"]}"`).

Reply:
This is actually a part of boilerplate code similar to this:
https://github.com/google/grr/blob/b472626b6ff4d05b24269a7ce52419d00a847607/grr/server/grr_response_server/bin/frontend.py#L325-L327
What do you think? Should we change both or stay consistent with the existing codebase?

Reply:
This linked code fragment was written way before we migrated to Python 3, so it still uses "the old way". In my opinion, new code should not repeat old mistakes for the sake of consistency. The linked fragment should be changed, I agree, but, since it is not related, not as part of this pull request.